File: src/app/_services/ingrediente-grupo.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable, signal } from '@angular/core';
import { take } from 'rxjs';
import { ToastrService } from 'ngx-toastr';
import { IngredienteGrupoDto } from '../_models/dtos';
import { PaginatedResult, setPaginationHeader } from '../../_helpers/pagination';
import { environment } from '../../_environment/environment';

@Injectable({ providedIn: 'root' })
export class IngredienteGrupoService {
  private http = inject(HttpClient);
  private toastr = inject(ToastrService);
  private baseUrl = `${environment.apiUrl}/ingredientegrupo`;

  grupos = signal<IngredienteGrupoDto[]>([]);
  paginatedResult = signal<PaginatedResult<IngredienteGrupoDto[]>>({});
  todos = signal<IngredienteGrupoDto[]>([]);

  retornaGrupos(pageNumber: number, pageSize: number, search = '', orderBy = 'Nome', order = 'asc') {
    const params = setPaginationHeader(pageNumber, pageSize, search, orderBy, order);
    return this.http.get<IngredienteGrupoDto[]>(this.baseUrl, { observe: 'response', params })
      .pipe(take(1))
      .subscribe({
        next: response => {
          this.grupos.set(response.body as IngredienteGrupoDto[]);
          this.paginatedResult.set({
            result: this.grupos(),
            pagination: JSON.parse(response.headers.get('Pagination')!)
          });
        }
      });
  }

  retornaTodos() {
    return this.http.get<IngredienteGrupoDto[]>(`${this.baseUrl}/todos`)
      .pipe(take(1))
      .subscribe({ next: lista => this.todos.set(lista) });
  }

  cadastrar(model: IngredienteGrupoDto) {
    return this.http.post<IngredienteGrupoDto>(this.baseUrl, model)
      .pipe(take(1))
      .subscribe({
        next: item => {
          this.grupos.update(l => [item, ...l]);
          this.toastr.success('Grupo cadastrado com sucesso');
        },
        error: err => this.toastr.error(err?.error || 'Erro ao cadastrar')
      });
  }

  atualizar(model: IngredienteGrupoDto) {
    return this.http.put<IngredienteGrupoDto>(this.baseUrl, model)
      .pipe(take(1))
      .subscribe({
        next: item => {
          this.grupos.update(l => l.map(m => m.id === item.id ? item : m));
          this.toastr.success('Grupo atualizado com sucesso');
        },
        error: err => this.toastr.error(err?.error || 'Erro ao atualizar')
      });
  }

  excluir(id: number) {
    return this.http.delete(`${this.baseUrl}/${id}`)
      .pipe(take(1))
      .subscribe({
        next: () => {
          this.grupos.update(l => l.filter(m => m.id !== id));
          this.toastr.success('Grupo removido');
        },
        error: err => this.toastr.error(err?.error || 'Erro ao excluir')
      });
  }
}

File: mobile/lib/_services/ingrediente_grupo_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import '../_helpers/api_helper.dart';
import '../_models/dtos.dart';

class IngredienteGrupoService extends ChangeNotifier {
  List<IngredienteGrupoDto> grupos = [];
  List<IngredienteGrupoDto> todos = [];
  PaginatedResult<IngredienteGrupoDto>? paginatedResult;
  bool carregando = false;

  Future<void> retornaGrupos(
    String token, {
    int pageNumber = 1,
    int pageSize = 10,
    String search = '',
  }) async {
    carregando = true;
    notifyListeners();
    try {
      final query = buildPaginationQuery(
        pageNumber: pageNumber,
        pageSize: pageSize,
        search: search,
        orderBy: 'ordem',
        order: 'asc',
      );
      final response = await http.get(
        Uri.parse('${baseUrl}ingredientegrupo?$query'),
        headers: buildHeaders(token),
      );
      if (response.statusCode == 200) {
        paginatedResult = parsePaginatedResponse(
          response,
          IngredienteGrupoDto.fromJson,
        );
        grupos = List.from(paginatedResult!.result);
      }
    } finally {
      carregando = false;
      notifyListeners();
    }
  }

  Future<void> retornaTodos(String token) async {
    try {
      final response = await http.get(
        Uri.parse('${baseUrl}ingredientegrupo/todos'),
        headers: buildHeaders(token),
      );
      if (response.statusCode == 200) {
        final List<dynamic> body = json.decode(response.body);
        todos = body.map((e) => IngredienteGrupoDto.fromJson(e)).toList();
        notifyListeners();
      }
    } catch (_) {}
  }

  Future<String?> cadastrar(IngredienteGrupoDto model, String token) async {
    try {
      final response = await http.post(
        Uri.parse('${baseUrl}ingredientegrupo'),
        headers: buildHeaders(token),
        body: json.encode(model.toJson()),
      );
      if (response.statusCode == 200) {
        final novo = IngredienteGrupoDto.fromJson(json.decode(response.body));
        grupos.insert(0, novo);
        notifyListeners();
        return null;
      }
      return 'Erro ao cadastrar grupo';
    } catch (e) {
      return 'Erro de conexão';
    }
  }

  Future<String?> atualizar(IngredienteGrupoDto model, String token) async {
    try {
      final response = await http.put(
        Uri.parse('${baseUrl}ingredientegrupo'),
        headers: buildHeaders(token),
        body: json.encode(model.toJson()),
      );
      if (response.statusCode == 200) {
        final atualizado = IngredienteGrupoDto.fromJson(
          json.decode(response.body),
        );
        final index = grupos.indexWhere((g) => g.id == atualizado.id);
        if (index != -1) grupos[index] = atualizado;
        notifyListeners();
        return null;
      }
      return 'Erro ao atualizar grupo';
    } catch (e) {
      return 'Erro de conexão';
    }
  }

  Future<String?> excluir(int id, String token) async {
    try {
      final response = await http.delete(
        Uri.parse('${baseUrl}ingredientegrupo/$id'),
        headers: buildHeaders(token),
      );
      if (response.statusCode == 204) {
        grupos.removeWhere((g) => g.id == id);
        notifyListeners();
        return null;
      }
      return 'Erro ao excluir grupo';
    } catch (e) {
      return 'Erro de conexão';
    }
  }
}
